### PR TITLE
refactor(module): update module

### LINKED ISSUE #12: test: write up test cases for errors
Currently, several `catch` clauses are ignored by the coverage tool. Test cases for them are not strictly necessary but would be appreciated ☺️

Link to code coverage: https://codecov.io/gh/nuxt-community/feed-module/src/master/lib/module.js

### LINKED ISSUE #40: unclear documentation
The top level configuration works as expected:

```
  modules: [
    ['@nuxtjs/feed', {
    }],
  ],
  feed: [
    {
     path: '/feed.xml',
     async create (feed) {},
     cacheTime: 1000 * 60 * 15,
     type: 'rss2',
     data: ['Some additional data']
    }
  ],
```

Moving this into the other section as the docs suggest:
```
  modules: [
    ['@nuxtjs/feed', {
        path: '/feed.xml',
        async create (feed) {},
        cacheTime: 1000 * 60 * 15,
        type: 'rss2',
        data: ['Some additional data']
    }],
  ],
```
gives
```
Cannot read property 'factory' of undefined                                                                                     17:19:11

  at ModuleContainer.feed (node_modules/@nuxtjs/feed/lib/module.js:24:27)
  at ModuleContainer.addModule (node_modules/@nuxt/core/dist/core.js:177:34)
  at promise.then (node_modules/@nuxt/utils/dist/utils.js:1664:43)
  at process.runNextTicks [as _tickCallback] (internal/process/next_tick.js:47:5)
  at Function.Module.runMain (internal/modules/cjs/loader.js:804:11)
  at internal/main/run_main_module.js:21:11
```

Checking the source and logging what `this.options` contains it seems like only the top level config is supported.

### LINKED ISSUE #41: feed output in static folder
Why is the plugin writing the feeds into the `static` folder?
I would expect the files to only show up in the `dist` folder.

### PR DESCRIPTION
**Pull request by ricardogobbosouza**
- Merge module options with feed top level
- ~~Feed output in generate folder~~ (NextPR)
- Add more tests
- Increase coverage

Resolves #12 
Resolves #39
Resolves #40
~~Resolves #41~~ (Next PR)

### CODE REVIEW

File: lib/module.js [hunk 1]
```diff
@@ -3,47 +3,31 @@ const { promisify } = require('util')
...
-
-  // Single feed
-  if (!Array.isArray(this.options.feed)) {
-    this.options.feed = [this.options.feed]
-  }
-
-  const options = Object.assign([], this.options.feed).map(o => Object.assign({}, defaults, o))
+async function feedModule(moduleOptions) {
+  const options = [
+    ...await parseOptions(moduleOptions),
```
Comment (manniL) [inline lib/module.js]: I wouldn't merge the options. Instead, check whether moduleOptions are populated. If so, use them. Else, use `this.options.feed`

File: lib/module.js [hunk 2]
```diff
@@ -3,47 +3,31 @@ const { promisify } = require('util')
...
+      createFeed(options[feedIndex], callback).catch(err => logger.error(err))
     }
   })
 
   feedCache.get = promisify(feedCache.get)
 
   options.forEach((feedOptions, index) => {
-    this.nuxt.hook('generate:before', async () => {
-      const xmlGeneratePath = resolve(this.options.srcDir, join('static', feedOptions.path))
+    this.nuxt.hook('generate:done', async () => {
```
Comment (manniL) [inline lib/module.js]: This is a breaking change

File: lib/module.js [hunk 3]
```diff
@@ -79,14 +89,21 @@ async function createFeed(feedOptions, callback) {
   }
 
   const feed = new Feed()
+  let result = ''
```
Comment (manniL) [inline lib/module.js]: Please avoid mutable variables where possible

File: lib/module.js (hunk 2 above)
Comment (ricardogobbosouza) [reply lib/module.js]: What do you recommend?

File: lib/module.js (hunk 1 above)
Comment (ricardogobbosouza) [reply lib/module.js]: I do not know if I agree, most modules mix, today is not even being used

File: lib/module.js (hunk 3 above)
Comment (ricardogobbosouza) [reply lib/module.js]: I'm going to change

File: lib/module.js (hunk 2 above)
Comment (manniL) [reply lib/module.js]: A separate PR for this change :wink:

File: lib/module.js (hunk 3 above)
Comment (ricardogobbosouza) [reply lib/module.js]: @manniL changed

File: lib/module.js (hunk 2 above)
Comment (ricardogobbosouza) [reply lib/module.js]: @manniL rollback, after approving this PR I will create a new one with this change

File: lib/module.js (hunk 1 above)
Comment (ricardogobbosouza) [reply lib/module.js]: @manniL ao you think we should concat instead of merge?

Comment (manniL) [reply lib/module.js]: @ricardogobbosouza it doesn't make sense for the user to use both options. :man_shrugging:

Comment (manniL) [reply lib/module.js]: That's why I removed the moduleOptions support for now.

I don't see a reason why we should add that for this module.

Comment (ricardogobbosouza) [reply lib/module.js]: @manniL 
Actually, the two together does not make sense, but so we can use or not.
I see no reason to change here

Comment (ricardogobbosouza) [reply lib/module.js]: Follow the same module https://github.com/nuxt-community/redirect-module/blob/master/lib/module.js#L2

File: test/module.test.js [hunk 4]
```diff
@@ -1,106 +1,229 @@
...
+
     const html = await get('/feed.xml')
     expect(html).toMatchSnapshot()
   })
 
   test('function rss', async () => {
-    nuxt = await setupNuxt(require('./fixture/configs/function_rss'))
+    nuxt = await setupNuxt({
+      ...config,
+      ...{
```
Comment (manniL) [inline test/module.test.js]: unnecessary spreading (same for other tests)

```js
{
  ...config,
  feed: () => createFeed()
}
```

Comment (ricardogobbosouza) [reply test/module.test.js]: @manniL In fact, it is necessary to cover the entire module

Comment (ricardogobbosouza) [reply test/module.test.js]: I maintain a configuration file and modify it with each test to increase coverage

File: lib/module.js [hunk 5]
```diff
@@ -3,47 +3,31 @@ const { promisify } = require('util')
...
-  if (!Array.isArray(this.options.feed)) {
-    this.options.feed = [this.options.feed]
-  }
-
-  const options = Object.assign([], this.options.feed).map(o => Object.assign({}, defaults, o))
+async function feedModule(moduleOptions) {
+  const options = [
+    ...await parseOptions(this.options.feed),
+    ...await parseOptions(moduleOptions)
+  ].map(o => ({ ...{}, ...defaults, ...o }))
```
Comment (manniL) [inline lib/module.js]:
```suggestion
  ].map(o => ({ ...defaults, ...o }))
```

File: test/module.test.js (hunk 4 above)
Comment (manniL) [reply test/module.test.js]: I don't think you got me right:

### What you do

```js
 ...config,
      ...{
        feed: [
          { ...createFeed(), ...{ path: '/feed1.xml' } },
          { ...createFeed(), ...{ path: '/feed2.xml' } }
        ]
      }
```

### What I suggested

```js
 ...config,
    feed: [
      { ...createFeed(), ...{ path: '/feed1.xml' } },
      { ...createFeed(), ...{ path: '/feed2.xml' } }
    ]
```

Comment (ricardogobbosouza) [reply test/module.test.js]: Sorry, I got it wrong, changed

**Comment by manniL** [comment]
Well done :relaxed:

**Comment by ricardogobbosouza** [comment]
:smile:

**Pull request merged**